### ISSUE TITLE
Env requirements

### ISSUE THREAD

**Issue by ZhengPeng7** [fanq15/GCoNet#7]
Could you provide the environment of your project?

I tried it many times, it worked very well on SoSOD3k and CoSal2015, but only got Emax around 0.73 on CoCA. I ran the latest code directly with default settings.

So I suspect a little about whether it's the **env requirements** that produced the gap.

Thanks a lot!

**Comment by fanq15** [comment]
Hi, you can try this conda env:
```
name: base
channels:
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - asn1crypto=1.2.0=py37_0
  - backcall=0.1.0=py37_0
  - beautifulsoup4=4.8.2=py37_0
  - blas=1.0=mkl
  - bzip2=1.0.8=h7b6447c_0
  - ca-certificates=2019.11.27=0
  - certifi=2019.11.28=py37_0
  - cffi=1.13.0=py37h2e261b9_0
  - chardet=3.0.4=py37_1003
  - conda=4.8.1=py37_0
  - conda-build=3.18.11=py37_0
  - conda-package-handling=1.6.0=py37h7b6447c_0
  - cryptography=2.8=py37h1ba5d50_0
  - cudatoolkit=10.1.243=h6bb024c_0
  - decorator=4.4.1=py_0
  - filelock=3.0.12=py_0
  - freetype=2.9.1=h8a8886c_1
  - glob2=0.7=py_0
  - idna=2.8=py37_0
  - intel-openmp=2019.4=243
  - ipython_genutils=0.2.0=py37_0
  - jedi=0.15.2=py37_0
  - jinja2=2.10.3=py_0
  - jpeg=9b=h024ee3a_2
  - libarchive=3.3.3=h5d8350f_5
  - libedit=3.1.20181209=hc058e9b_0
  - libffi=3.2.1=hd88cf55_4
  - libgcc-ng=9.1.0=hdf63c60_0
  - libgfortran-ng=7.3.0=hdf63c60_0
  - liblief=0.9.0=h7725739_2
  - libpng=1.6.37=hbc83047_0
  - libstdcxx-ng=9.1.0=hdf63c60_0
  - libtiff=4.1.0=h2733197_0
  - libxml2=2.9.9=hea5a465_1
  - lz4-c=1.8.1.2=h14c3975_0
  - lzo=2.10=h49e0be7_2
  - markupsafe=1.1.1=py37h7b6447c_0
  - mkl=2019.4=243
  - mkl-service=2.3.0=py37he904b0f_0
  - mkl_fft=1.0.15=py37ha843d7b_0
  - mkl_random=1.1.0=py37hd6b4f25_0
  - ncurses=6.1=he6710b0_1
  - olefile=0.46=py37_0
  - openssl=1.1.1d=h7b6447c_3
  - parso=0.5.2=py_0
  - patchelf=0.10=he6710b0_0
  - pexpect=4.7.0=py37_0
  - pickleshare=0.7.5=py37_0
  - pkginfo=1.5.0.1=py37_0
  - psutil=5.6.7=py37h7b6447c_0
  - ptyprocess=0.6.0=py37_0
  - py-lief=0.9.0=py37h7725739_2
  - pycosat=0.6.3=py37h14c3975_0
  - pycparser=2.19=py37_0
  - pygments=2.5.2=py_0
  - pyopenssl=19.0.0=py37_0
  - pysocks=1.7.1=py37_0
  - python=3.7.4=h265db76_1
  - python-libarchive-c=2.8=py37_13
  - pytz=2019.3=py_0
  - pyyaml=5.2=py37h7b6447c_0
  - readline=7.0=h7b6447c_5
  - requests=2.22.0=py37_0
  - ripgrep=11.0.2=he32d670_0
  - ruamel_yaml=0.15.46=py37h14c3975_0
  - setuptools=41.4.0=py37_0
  - six=1.12.0=py37_0
  - soupsieve=1.9.5=py37_0
  - sqlite=3.30.0=h7b6447c_0
  - tk=8.6.8=hbc83047_0
  - tqdm=4.36.1=py_0
  - traitlets=4.3.3=py37_0
  - urllib3=1.24.2=py37_0
  - wcwidth=0.1.7=py37_0
  - wheel=0.33.6=py37_0
  - xz=5.2.4=h14c3975_4
  - yaml=0.1.7=had09818_2
  - zlib=1.2.11=h7b6447c_3
  - zstd=1.3.7=h0b5b093_0
  - pip:
    - absl-py==0.9.0
    - addict==2.2.1
    - albumentations==0.4.5
    - argparse==1.4.0
    - attrs==19.3.0
    - bleach==3.1.3
    - boto3==1.12.26
    - botocore==1.15.26
    - cachetools==4.0.0
    - cloudpickle==1.4.1
    - cycler==0.10.0
    - cython==0.29.19
    - dataclasses==0.6
    - defusedxml==0.6.0
    - detectron2==0.3+cu101
    - docutils==0.15.2
    - easydict==1.9
    - entrypoints==0.3
    - flake8==3.8.3
    - flake8-import-order==0.18.1
    - futures==3.1.1
    - fvcore==0.1.2.post20201204
    - gensim==3.8.1
    - google-api-core==1.16.0
    - google-auth==1.11.3
    - google-auth-oauthlib==0.4.1
    - google-cloud-core==1.3.0
    - google-cloud-storage==1.26.0
    - google-resumable-media==0.5.0
    - googleapis-common-protos==1.51.0
    - grpcio==1.29.0
    - h5py==2.10.0
    - imageio==2.8.0
    - imagesize==1.2.0
    - imgaug==0.2.6
    - imgcat==0.5.0
    - importlib-metadata==1.5.0
    - iniconfig==1.0.1
    - install==1.3.4
    - iopath==0.1.8
    - ipdb==0.13.3
    - ipykernel==5.2.0
    - ipython==5.7.0
    - ipywidgets==7.5.1
    - jieba==0.42.1
    - jmespath==0.9.5
    - joblib==0.14.1
    - jsonschema==3.2.0
    - jupyter==1.0.0
    - jupyter-client==6.1.0
    - jupyter-console==6.1.0
    - jupyter-core==4.6.3
    - kiwisolver==1.1.0
    - llvmlite==0.33.0
    - lmdb==0.98
    - markdown==3.2.2
    - matplotlib==3.2.1
    - mccabe==0.6.1
    - mistune==0.8.4
    - mmcv==1.2.1
    - mmcv-full==1.2.1
    - mock==4.0.2
    - more-itertools==8.4.0
    - moviepy==0.2.3.5
    - multiscaledeformableattention==1.0
    - mxboard==0.1.0
    - natsort==7.0.1
    - nbconvert==5.6.1
    - nbformat==5.0.4
    - networkx==2.4
    - ninja==1.9.0.post1
    - nltk==3.4.5
    - nose==1.3.7
    - notebook==6.0.3
    - numba==0.50.1
    - numpy==1.18.4
    - oauthlib==3.1.0
    - opencv-python==4.2.0.32
    - packaging==20.4
    - pandas==1.0.3
    - pandocfilters==1.4.2
    - pillow==8.0.1
    - pip==20.1.1
    - pluggy==0.13.1
    - polygon3==3.0.8
    - prometheus-client==0.7.1
    - prompt-toolkit==1.0.18
    - protobuf==3.11.3
    - py==1.9.0
    - py-cpuinfo==7.0.0
    - pyasn1==0.4.8
    - pyasn1-modules==0.2.8
    - pycocotools==2.0.2
    - pycodestyle==2.6.0
    - pydot==1.4.1
    - pyflakes==2.2.0
    - pyparsing==2.4.6
    - pyrsistent==0.15.7
    - pytest==6.0.1
    - pytest-benchmark==3.2.3
    - python-dateutil==2.8.1
    - python-levenshtein==0.12.0
    - pywavelets==1.1.1
    - pyzmq==19.0.0
    - qtconsole==4.7.1
    - qtpy==1.9.0
    - requests-oauthlib==1.3.0
    - rsa==4.0
    - s3transfer==0.3.3
    - scikit-image==0.17.2
    - scikit-learn==0.22.2.post1
    - scipy==1.4.1
    - script-utils==0.0.1
    - seaborn==0.10.0
    - send2trash==1.5.0
    - shapely==1.7.0
    - simplegeneric==0.8.1
    - smart-open==1.10.0
    - tabulate==0.8.6
    - tao==0.1.0
    - tensorboard==2.2.2
    - tensorboard-plugin-wit==1.6.0.post3
    - tensorboardx==2.0
    - termcolor==1.1.0
    - terminado==0.8.3
    - testpath==0.4.4
    - tifffile==2020.5.30
    - toml==0.10.1
    - torch==1.7.0
    - torch-encoding==1.2.2b20210516
    - torchaudio==0.7.0
    - torchvision==0.8.1
    - tornado==6.0.4
    - typing-extensions==3.7.4.3
    - webencodings==0.5.1
    - werkzeug==1.0.1
    - widgetsnbextension==3.5.1
    - xgboost==1.0.2
    - xmltodict==0.12.0
    - yacs==0.1.7
    - yapf==0.30.0
    - youtube-dl==2020.7.28
    - zipp==3.1.0
prefix: /opt/conda

```

**Comment by fanq15** [comment]
@ZhengPeng7 Hi, I found the reason for the performance gap on CoCA. It is not caused by the env setting. It is caused by the different group pairs in the training stage. The group pair is very important for Collaborative training, especially for the challenging CoCA dataset. If the group pairs are very different classes, the collaborative training can not separate similar classes.I will publish the training pairs later to reproduce the performance on CoCA. And it is also a promising research direction to mine effective group pairs in collaborative training.

**Comment by ZhengPeng7** [comment]
Thanks a lot!👍 That's really an interesting topic or at least a trick worth some exploration.

**Comment by fanq15** [comment]
Hi, please try the new code with deterministic input sequence and data augmentation. You can obtain similar performance with this code.

**Comment by fanq15** [comment]
@ZhengPeng7 When I train the model multiple times, I found that I can usually get better performance on Cosal2015 and CoSOD3k datasets and worse performance on the CoCA dataset.

**Comment by ZhengPeng7** [comment]
> @ZhengPeng7 When I train the model multiple times, I found that I can usually get better performance on Cosal2015 and CoSOD3k datasets and worse performance on the CoCA dataset.

Yeah, my Emax could achieve **0.863** on _CoSOD3k_ and **0.891** on _CoSal2015_.
The performance was very stable on these two datasets.

Thanks for your effort!

**Comment by ldkong1205** [comment]
> @ZhengPeng7 When I train the model multiple times, I found that I can usually get better performance on Cosal2015 and CoSOD3k datasets and worse performance on the CoCA dataset.

Does it possibly be caused by the training-testing inconsistency? I noticed that the training samples in DUTS are similar to that of CoSal2015 and CoSOD3k, but different from samples in CoCA. Generally speaking, most samples in DUTS, CoSal2015, and CoSOD3k have a uni-object distribution, while samples in CoCA usually consist of multiple foreground objects.

**Comment by ZhengPeng7** [comment]
> > @ZhengPeng7 When I train the model multiple times, I found that I can usually get better performance on Cosal2015 and CoSOD3k datasets and worse performance on the CoCA dataset.
> 
> Does it possibly be caused by the training-testing inconsistency? I noticed that the training samples in DUTS are similar to that of CoSal2015 and CoSOD3k, but different from samples in CoCA. Generally speaking, most samples in DUTS, CoSal2015, and CoSOD3k have a uni-object distribution, while samples in CoCA usually consist of multiple foreground objects.

Nice explanation for this phenomenon, I will try to find and show some typical fail cases to see whether it is the problem.

**Comment by ldkong1205** [comment]
> > > @ZhengPeng7 When I train the model multiple times, I found that I can usually get better performance on Cosal2015 and CoSOD3k datasets and worse performance on the CoCA dataset.
> > 
> > 
> > Does it possibly be caused by the training-testing inconsistency? I noticed that the training samples in DUTS are similar to that of CoSal2015 and CoSOD3k, but different from samples in CoCA. Generally speaking, most samples in DUTS, CoSal2015, and CoSOD3k have a uni-object distribution, while samples in CoCA usually consist of multiple foreground objects.
> 
> Nice explanation for this phenomenon, I will try to find and show some typical fail cases to see whether it is the problem.

👍

**Comment by ZhengPeng7** [comment]
## CoCA
I think it's the ability to collaboration that made the performance decreased a lot on **CoCA**-- sometimes the model didn't manage to learn it.
For example, in the accordion class, the salmaps generated by the best model offered in the original paper are **much better** than those generated by an arbitrarily retrained model, in terms of **choosing the co-class**.

![CoCA_comp](https://user-images.githubusercontent.com/25921713/119486916-d2283c00-bd69-11eb-858c-c0f3f6bb894d.png)

+ In CoCA, there are too many distracting classes in it, especially persons. If there are not many distractors, these two models performed similarly to each other:
![CoCA_comp_headphone](https://user-images.githubusercontent.com/25921713/119487582-a0fc3b80-bd6a-11eb-92d5-567dfc301213.png)


## CoSal2015
Take CoSal2015 for the comparison, you can see that the problem above also occurred here, but not as frequently as it did in CoCA.
![CoSal_comp_bad](https://user-images.githubusercontent.com/25921713/119487689-c12bfa80-bd6a-11eb-8778-ca2d4cece79f.png)

Take the spider as an example.
![CoSal_comp_ladybird](https://user-images.githubusercontent.com/25921713/119487318-495dd000-bd6a-11eb-8f56-a0b18bae0ed0.png)

## In conclusion:
1. All the models still need improvement on segmenting the just complete object.
2. Extra improvement should be given to help models accurately **localize co-objects**.
